### PR TITLE
Inference Phase Two

### PR DESCRIPTION
**Pull request by Nargeshdb**
This PR infers:

- `@NotOwning` for return types.
- `@MustCallAlias` for both return types and parameter types.
- `@Owning` for constructor parameters when there are multiple owning fields.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java [hunk 1]
```diff
@@ -248,7 +248,8 @@ public boolean isValidUse(
           useType.hasPrimaryAnnotation(PolyMustCall.class)
               && !AnnotationUtils.containsSameByClass(
                   elt.getAnnotationMirrors(), PolyMustCall.class);
-      if (ajavaFileHasMustCallAlias) {
+      if (ajavaFileHasMustCallAlias
+          || AnnotationUtils.containsSameByClass(elt.getAnnotationMirrors(), MustCallAlias.class)) {
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:252]: Why is this change necessary? It seems redundant with line 237 above.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 2]
```diff
@@ -128,6 +141,13 @@ public class MustCallInference {
   /** The element referring to the enclosing class of the current method. */
   private final @Nullable TypeElement classElt;
 
+  /**
+   * This map is used to track alias relationships between return nodes and method parameters. The
+   * keys are the obligation of return nodes, and the values are the index of current method formal
+   * parameter (0-based) that is alias with the return node.
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:147]: WPI has switched to 1-based indexing for the parameters (with 0 representing the receiver), and I think we should do the same here for consistency.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 3]
```diff
@@ -226,10 +252,43 @@ private void runInference() {
     addMemberAndClassAnnotations();
   }
 
+  /**
+   * Analyzes a return statement and performs two computations:
+   *
+   * <ul>
+   *   <li>If the return node is a field with non-empty must-call obligations, adds a {@link
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:259]: To clarify this, I suggest changing "return node" to "the returned expression"

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 4]
```diff
@@ -226,10 +252,43 @@ private void runInference() {
...
+   *       NotOwning} annotation to the return type of the current method.
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:260]: Shouldn't this only occur if, in a given method, *every* returned expression has this property? Do we have a test for a method that returns a field in one place, and return something else (e.g., something with an obligation) at a different return site? We should add such a test if not.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 5]
```diff
@@ -367,41 +438,65 @@ private void analyzeOwnershipTransferAtAssignment(
       if (!TreeUtils.isConstructor(methodTree)) {
         disposedFields.remove((VariableElement) lhsElement);
       }
-      addOwningToParamsIfDisposedAtAssignment(obligations, rhsObligation, rhs);
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:370]: Why were the two calls to this method (here and at line 372 below) removed? These changes don't seem necessary (maybe they are unintentional? the only parts of these changes that seem required are the new lines 447 to 453, based on the Javadoc changes)

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 6]
```diff
@@ -670,11 +767,46 @@ private void computeOwningFromInvocation(Set<Obligation> obligations, Node invoc
...
     } else if (invocation instanceof MethodInvocationNode) {
+      computeMustCallAliasFromThisOrSuperCall(obligations, (MethodInvocationNode) invocation);
       inferOwningParamsViaOwnershipTransfer(obligations, invocation);
       computeOwningForArgsOrReceiverOfCall(obligations, (MethodInvocationNode) invocation);
     }
   }
 
+  /**
+   * Adds @MustCallAlias annotation to a parameter of the current method when it is passed in
+   * a @MustCallAlias position in a {@literal this} or {@literal super} call.
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:778]: Add the word "constructor" in this comment somewhere, because right now I think it's a bit confusingly worded. In particular, it's possible to have `super` calls that are not to the superclass constructor: for example, to access an overridden method.

**Review by kelloggm — CHANGES_REQUESTED** [review]
Overall this PR looks much simpler than phase 1, thanks! I think it's mostly in good shape.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java (hunk 5 above)
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:370]: This change was intentional. I changed `addOwningToParamsIfDisposedAtAssignment` method to `getIndexOfParam` to reduce the redundant computations in this method and also in the `analyzeReturnNode`. The new method returns the index of formal parameter that is alias with the given obligation. In this method, the given obligation is the rhs obligation, and in the  `analyzeReturnNode` is the obligation of the return node.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java (hunk 4 above)
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:260]: Based on the algorithm we don't need to check all return statements but if you think it might be a common pattern I can change this part.

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:260]: Sorry, I think I misremembered the algorithm here. I checked the paper, and you are correct. Let's add a test for this situation anyway (i.e., a getter that sometimes returns a field and sometimes returns something else, like `null`) and make sure `@NotOwning` is inferred (the code here should do it, but it's good to have the regression test).

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 7]
```diff
@@ -241,6 +300,17 @@ private void addMemberAndClassAnnotations() {
       addEnsuresCalledMethods();
     }
 
+    // If all return statements alias the same parameter index, then add the @MustCallAlias
+    // annotation to that parameter.
+    if (!returnNodeToParameterIndexMap.isEmpty()) {
+      if (returnNodeToParameterIndexMap.values().stream().distinct().count() == 1) {
+        int indexOfParam = returnNodeToParameterIndexMap.values().iterator().next();
+        if (indexOfParam > 0) {
```
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:308]: Why not permit inference of `@MustCallAlias` on the receiver? I think that could be a sensible specification.

If there is a technical reason (I suspect there probably is!), please add a comment explaining it.

**Review by kelloggm — CHANGES_REQUESTED** [review]
This is close to being ready to merge, but I think some clarifying comments are necessary still.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java [hunk 8]
```diff
@@ -368,41 +441,65 @@ private void analyzeOwnershipTransferAtAssignment(
...
       LocalVariableNode lhsVar = (LocalVariableNode) lhs;
       mcca.updateObligationsForPseudoAssignment(obligations, assignmentNode, lhsVar, rhs);
     }
   }
 
   /**
-   * If a must-call obligation of some alias of method parameter p is satisfied during the
-   * assignment, add an @Owning annotation to p, and remove the rhs node from the obligations set,
-   * since it no longer needs to be tracked.
+   * Return the index of the method parameter that is an alias of the given {@code obligation}, if
```
[the remark this answers is not in the document]
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: It wasn't 1-based but I changed it to be 1-based to make it cleaner.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java (hunk 7 above)
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: We didn't have a rule for the receiver type and I didn't consider this case, but I added some changes to support it. Thanks for bringing this to my attention.

**Comment by msridhar** [comment]
Closing as this is all merged upstream now
